### ISSUE TITLE
Write default implementation to getLogAction via logActionL

### ISSUE THREAD

**Issue by chshersh** [kowainik/co-log#123]
So it will be enough to implement either `logActionL` or `get + set`.

**Comment by SanchayanMaity** [comment]
@chshersh Is this something I can work on if my understanding is correct?

I see *logActionL* is a Lens. If I understand correctly, the *get* and *set* you mention refer to the getter and setter in the context of Lens? Depending on what operation I apply using logActionL, I can retrieve *getLogAction* and *setLogAction* from logActionL. Am I thinking correctly here?

**Comment by chshersh** [comment]
@SanchayanMaity The `HasLog` typeclass is exactly about getting and setting `LogAction` in some `env`ironment data type. Lens is a smart object that can be used as a getter and a setter. If you're interested in understanding lens concept better, I can recommend the following blog post by @vrom911:

* [Write yourself a lens](https://vrom911.github.io/blog/write-yourself-a-lens)

Since lens has getter and setter capabilities, it should be enough to define only `logActionL` lens in the `HasLog` typeclass. However, currently this is not the case, because `getLogAction` function doesn't have the default implementation, so you need to implement it anyway which seems redundant. 

So, in simple words, this issue is about the following:

1. Add default implementation of `getLogAction`.
2. Change _minimal complete definition_ accordingly:

https://github.com/kowainik/co-log/blob/434ee55f12d839e9c23909107d07c8866f0bfb7e/co-log-core/src/Colog/Core/Class.hs#L39-L40

**Comment by SanchayanMaity** [comment]
@chshersh Does this look good?

```haskell
diff --git a/co-log-core/src/Colog/Core/Class.hs b/co-log-core/src/Colog/Core/Class.hs
index d2f6c92..1291f1c 100644
--- a/co-log-core/src/Colog/Core/Class.hs
+++ b/co-log-core/src/Colog/Core/Class.hs
@@ -19,7 +19,7 @@ module Colog.Core.Class
        ) where
 
 import Colog.Core.Action (LogAction)
-
+import Data.Functor.Const
 
 -- to inline lens better
 {- HLINT ignore "Redundant lambda" -}
@@ -37,10 +37,11 @@ Every instance of the this typeclass should satisfy the following laws:
 4. __Set-Over:__ @'overLogAction' f env ≡ 'setLogAction' (f $ 'getLogAction' env) env@
 -}
 class HasLog env msg m where
-    {-# MINIMAL getLogAction, (setLogAction | overLogAction) #-}
+    {-# MINIMAL logActionL , (setLogAction | overLogAction) #-}
 
     -- | Extracts 'LogAction' from the environment.
     getLogAction :: env -> LogAction m msg
+    getLogAction = getConst . logActionL Const
 
     -- | Sets 'LogAction' to the given one inside the environment.
     setLogAction :: LogAction m msg -> env -> env
```

**Comment by chshersh** [comment]
@SanchayanMaity A few notes:

1. Imports must have explicit lists
2. MINIMAL pragma doesn't look right. Now it says: `logActionL` and (`setLogAction` or `overLogAction`).

**Comment by SanchayanMaity** [comment]
@chshersh Thank you for your patience & sorry for asking stupid questions.

> @SanchayanMaity A few notes:
> 
>     1. Imports must have explicit lists

If I do an import with explicit lists as below which is what I was trying to do
```haskell
import Data.Functor.Const (Const, getConst)
```
I get an error, which I do not expect
```haskellco-log-core     > /home/sanchayan/GitSources/co-log/co-log-core/src/Colog/Core/Class.hs:44:42: error:
co-log-core     >     • Data constructor not in scope:
co-log-core     >         Const
co-log-core     >           :: LogAction m0 msg0 -> Const (LogAction m msg) (LogAction m0 msg0)
co-log-core     >     • Perhaps you meant variable ‘const’ (imported from Prelude)
co-log-core     >       Perhaps you want to add ‘Const’ to the import list in the import of
co-log-core     >       ‘Data.Functor.Const’ (src/Colog/Core/Class.hs:22:1-43).
co-log-core     >    |
co-log-core     > 44 |     getLogAction = getConst . logActionL Const
co-log-core     >    |                                          ^^^^^
```


> 
>     2. MINIMAL pragma doesn't look right. Now it says: `logActionL` and (`setLogAction` or `overLogAction`).

Either someone provides logActionL and we retrieve getLogAction from that or they provide getLogAction. Ditto for the other one. So below?

```haskell
{-# MINIMAL (getLogAction | logActionL) , (setLogAction | overLogAction) #-}
```

**Comment by chshersh** [comment]
@SanchayanMaity you need to import constructor of the `Const` data type. You can import everything you need for this issue by using the following import:

```haskell
import Data.Const (Const (..))
```

>  So below?

Below doesn't look correct either. `logActionL` should be enough to implement the instance. So the pragma should have the shape: `logActionL` or _smth else_. Where _smth else_ is what was before.

**Comment by vrom911** [comment]
Closed in #153
